src/pybind/cephfs/cephfs.pyx

```diff
@@ -2832,3 +2837,148 @@ cdef class LibCephFS(object):
+                    self.set_readdir_error()
+
+            def add_to_de_ignore_list(self, d_name):
+                self.de_ignore_list.append(d_name)
+
+            def set_readdir_error(self):
+                self.has_readdir_failed = True
+
+            def should_skip_d_name(self, d_name):
+                return d_name in self.de_ignore_list
+
+            def has_any_fs_op_failed(self):
+                return self.has_readdir_failed or len(self.de_ignore_list) > 0
+
+        # stack needed for traversing file heirarchy under trash_path in
+        # depth-first, non-recursive fashion. each stack member is an
+        # instance of class Dir (declared above, inside this method).
+        stack = deque([Dir(b".", trash_path, self),])
+
+        while stack and not should_cancel():
+            curr_dir = stack[-1]
+
+            # de = directory entry
+            de = curr_dir.read_dir()
+            while de and not should_cancel():
+                # if dir entry has been removed, rewinddir() should be
+                # called since POSIX doesn't guarantee anything regarding
+                # behaviour of readdir() when readdir() and unlink()/
+                # rmdir() calls are interleaved. Whenever calls to unlink()/
+                # rmdir() are made, reading dir needs to be restarted to
+                # check if the dir is actually empty.
+                de_has_been_removed = False
+
+                de_path = os.path.join(trash_path, curr_dir.path, de.d_name)
+
+                if de.is_dir():
+                    try:
+                        self.rmdir(de_path)
+                        de_has_been_removed = True
+                    except ObjectNotEmpty:
+                        try:
+                            stack.append(
+                                Dir(os.path.join(curr_dir.path, de.d_name),
+                                trash_path, self))
+                            break
+                        except Error as e:
+                            log.error('Exception occured during opendir(): '
+                                      f'"{e}"')
+                            curr_dir.add_to_de_ignore_list(de.d_name)
+                            continue
+                    except Error as e:
+                        log.error('Exception occured during rmdir(): '
+                                  f'"{e}"')
+                        curr_dir.add_to_de_ignore_list(de.d_name)
+                        continue
+                else:
+                    try:
+                        self.unlink(de_path)
+                        de_has_been_removed = True
+                    except Error as e:
+                        log.error('Exception occured during unlink(): '
+                                  f'"{e}"')
+                        curr_dir.add_to_de_ignore_list(de.d_name)
+                        continue
+
+                if de_has_been_removed:
+                    curr_dir.handle.rewinddir()
```

Comment:
Only do this after completing iteration. `unlink` does not invalidate the previously fetched contents of `getdents` for `readdir`.

The important point here is that `unlink` may cause the directory contents to reorder.